feat: add C and Fortran implementation for `blas/base/zdscal`

ShabiShett07 opened the pull request:
Progresses #2039.

## Description

> What is the purpose of this pull request?

This pull request:

-   add C and Fortran implementation for `blas/base/zdscal`

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   progresses #2039

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

ShabiShett07 committed:
feat: add C implementation for blas/base/zdscal

ShabiShett07 committed:
chore: add implementation

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: missing_dependencies
  - task: lint_c_benchmarks
    status: missing_dependencies
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

ShabiShett07 committed:
chore: add implementation

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

ShabiShett07 commented:
/stdlib update-copyright-years

stdlib-bot committed:
chore: update copyright years

kgryte commented on lib/node_modules/@stdlib/blas/base/zdscal/README.md:
````diff
@@ -149,6 +149,136 @@ console.log( zx.toString() );
+```c
+#include "stdlib/complex/float64/ctor.h"
+
+stdlib_complex128_t x[] = { stdlib_complex128( 1.0, 2.0 ), stdlib_complex128( 3.0, 4.0 ), stdlib_complex128( 5.0, 6.0 ) }; // interleaved real and imaginary components
````
```suggestion
stdlib_complex128_t x[] = {
	stdlib_complex128( 1.0, 2.0 ),
	stdlib_complex128( 3.0, 4.0 ),
	stdlib_complex128( 5.0, 6.0 )
};
```

kgryte commented on lib/node_modules/@stdlib/blas/base/zdscal/README.md:
````diff
@@ -149,6 +149,136 @@ console.log( zx.toString() );
+Scales a double-precision complex floating-point vector by a double-precision floating-point constant using alternative indexing semantics.
+
+```c
+stdlib_complex128_t x[] = { stdlib_complex128( 1.0, 2.0 ), stdlib_complex128( 3.0, 4.0 ), stdlib_complex128( 5.0, 6.0 ) }; // interleaved real and imaginary components
````
```suggestion
#include "stdlib/complex/float64/ctor.h"

stdlib_complex128_t x[] = {
	stdlib_complex128( 1.0, 2.0 ),
	stdlib_complex128( 3.0, 4.0 ),
	stdlib_complex128( 5.0, 6.0 )
};
```

kgryte commented on lib/node_modules/@stdlib/blas/base/zdscal/README.md:
```diff
@@ -149,6 +149,136 @@ console.log( zx.toString() );
+#include <stdio.h>
+
+int main( void ) {
+    // Create strided arrays:
+    stdlib_complex128_t x[] = { stdlib_complex128( 1.0, 2.0 ), stdlib_complex128( 3.0, 4.0 ), stdlib_complex128( 5.0, 6.0 ), stdlib_complex128( 7.0, 8.0 ) };
```
```suggestion
    // Create a strided array:
    stdlib_complex128_t x[] = {
    	stdlib_complex128( 1.0, 2.0 ),
    	stdlib_complex128( 3.0, 4.0 ),
    	stdlib_complex128( 5.0, 6.0 ),
    	stdlib_complex128( 7.0, 8.0 )
    };
```

kgryte commented on lib/node_modules/@stdlib/blas/base/zdscal/benchmark/benchmark.native.js:
```diff
@@ -0,0 +1,113 @@
+* @returns {Function} benchmark function
+*/
+function createBenchmark( len ) {
+	var x;
+
+	x = new Complex128Array( uniform( len*2, -100.0, 100.0, options ) );
+
```
```suggestion
	var x = new Complex128Array( uniform( len*2, -100.0, 100.0, options ) );
```

kgryte commented on lib/node_modules/@stdlib/blas/base/zdscal/benchmark/benchmark.ndarray.native.js:
```diff
@@ -0,0 +1,113 @@
+	var x;
+
+	x = new Complex128Array( uniform( len*2, -100.0, 100.0, options ) );
+
```
```suggestion
	var x = new Complex128Array( uniform( len*2, -100.0, 100.0, options ) );
```

kgryte commented on lib/node_modules/@stdlib/blas/base/zdscal/benchmark/c/benchmark.length.c:
```diff
@@ -0,0 +1,192 @@
+	int i;
+
+	for ( i = 0; i < len; i++ ) {
+		x[ i ] = stdlib_complex128( ( stdlib_complex128_real( rand_complex_128() ) *10000.0 ) - 5000.0, ( stdlib_complex128_imag( rand_complex_128() ) *10000.0 ) - 5000.0 ) ;
```
This is an inefficient way to do this. Refactor back to using `rand_double` and then do

```
x[ i ] = stdlib_complex128( (rand_double()*10000.0)-5000.0, (rand_double()*10000.0)-5000.0 );
```

kgryte commented on lib/node_modules/@stdlib/blas/base/zdscal/benchmark/c/benchmark.length.c:
```diff
@@ -0,0 +1,192 @@
+	t = tic();
+	for ( i = 0; i < iterations; i++ ) {
+		c_zdscal( len, 2.0, x, 1 );
+		if ( stdlib_base_is_nan( stdlib_complex128_real( x[ 0 ] ) ) ) {
```
```suggestion
		if ( stdlib_base_is_nan( stdlib_complex128_real( x[ i%len ] ) ) ) {
```

kgryte commented on lib/node_modules/@stdlib/blas/base/zdscal/benchmark/c/benchmark.length.c:
```diff
@@ -0,0 +1,192 @@
+		}
+	}
+	elapsed = tic() - t;
+	if ( stdlib_base_is_nan( stdlib_complex128_real( x[ 0 ] ) ) ) {
```
```suggestion
	if ( stdlib_base_is_nan( stdlib_complex128_imag( x[ i%len ] ) ) ) {
```

kgryte commented on lib/node_modules/@stdlib/blas/base/zdscal/benchmark/c/benchmark.length.c:
```diff
@@ -0,0 +1,192 @@
+	int i;
+
+	for ( i = 0; i < len; i++ ) {
+		x[ i ] = stdlib_complex128( ( stdlib_complex128_real( rand_complex_128() ) *10000.0 ) - 5000.0, ( stdlib_complex128_imag( rand_complex_128() ) *10000.0 ) - 5000.0 ) ;
```
Same comments and changes as above.

kgryte commented on lib/node_modules/@stdlib/blas/base/zdscal/examples/c/example.c:
```diff
@@ -0,0 +1,48 @@
+
+int main( void ) {
+	// Create strided arrays:
+	stdlib_complex128_t x[] = { stdlib_complex128( 1.0, 2.0 ), stdlib_complex128( 3.0, 4.0 ), stdlib_complex128( 5.0, 6.0 ), stdlib_complex128( 7.0, 8.0 ) };
```
Same changes as README.

kgryte commented on lib/node_modules/@stdlib/blas/base/zdscal/include/stdlib/blas/base/zdscal.h:
```diff
@@ -0,0 +1,49 @@
+/**
+* Scales a double-precision complex floating-point vector by a double-precision floating-point constant.
+*/
+void API_SUFFIX(c_zdscal)( const CBLAS_INT N, const double alpha, stdlib_complex128_t *X, const CBLAS_INT strideX );
```
```suggestion
void API_SUFFIX(c_zdscal)( const CBLAS_INT N, const double alpha, void *X, const CBLAS_INT strideX );
```

You need to fix this everywhere. CBLAS convention is for complex number arrays to be `void *`.

kgryte replied:
Including README and all other header files.

kgryte replied:
...and source files.

kgryte commented on lib/node_modules/@stdlib/blas/base/zdscal/manifest.json:
```diff
@@ -0,0 +1,322 @@
+    {
+      "task": "build",
+      "os": "win",
+      "blas": "openblas",
```
Windows should not have OpenBLAS configurations. See https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/blas/base/ccopy/manifest.json for reference. I suggest also ordering the configurations based on os, as done in `ccopy`.

kgryte commented on lib/node_modules/@stdlib/blas/base/zdscal/manifest.json:
```diff
@@ -0,0 +1,322 @@
+      "dependencies": [
+        "@stdlib/blas/base/shared"
+      ]
+    },
```
Where are the normal mac build configurations? Again, see `ccopy`. It is not clear how or why you specified the set of configurations you did. You should have good reason for deviating from prior art and these deviations should be explicitly described in your pull request description. Seeing things like this just decreases confidence that the changes you are wanting to introduce are actually valid and well-reasoned, thus decreasing the likelihood that your contributions will be reviewed and ultimately merged.

kgryte commented on lib/node_modules/@stdlib/blas/base/zdscal/manifest.json:
```diff
@@ -0,0 +1,322 @@
+        "@stdlib/complex/float64/imag",
+        "@stdlib/complex/float64/base/scale"
+      ]
+    },
```
Why does Linux not have OpenBLAS configurations?

kgryte commented on lib/node_modules/@stdlib/blas/base/zdscal/src/addon.c:
```diff
@@ -0,0 +1,65 @@
+	STDLIB_NAPI_ARGV_INT64( env, strideX, argv, 3 );
+	STDLIB_NAPI_ARGV_DOUBLE( env, alpha, argv, 1 );
+	STDLIB_NAPI_ARGV_STRIDED_COMPLEX128ARRAY( env, X, N, strideX, argv, 2 );
+	stdlib_complex128_t *x = (stdlib_complex128_t *)X;
+	API_SUFFIX(c_zdscal)( N, alpha, x, strideX );
```
This isn't correct. You should be doing

```suggestion
	API_SUFFIX(c_zdscal)( N, alpha, (void *)X, strideX );
```

Again, why are you deviating from existing prior art such as https://github.com/stdlib-js/stdlib/blob/bab1647c1309043335a7ec9badb8f3e629bf9136/lib/node_modules/%40stdlib/blas/base/ccopy/src/addon.c#L41? You need to explain your decision-making in your OP.

kgryte replied:
It is apparent that you are lacking context. Why `void *`? That is because we follow CBLAS. As such, any user familiar with BLAS who attempts to use our C implementations will be expecting that they should perform a `void *)` cast. Us using a custom stdlib type doesn't match that expectation. There is historical reasons for using `void *`, namely platforms did not always have a standardized complex number representation.

kgryte commented on lib/node_modules/@stdlib/blas/base/zdscal/src/addon.c:
```diff
@@ -0,0 +1,65 @@
+	STDLIB_NAPI_ARGV_INT64( env, offsetX, argv, 4 );
+	STDLIB_NAPI_ARGV_DOUBLE( env, alpha, argv, 1 );
+	STDLIB_NAPI_ARGV_STRIDED_COMPLEX128ARRAY( env, X, N, strideX, argv, 2 );
+	stdlib_complex128_t *x = (stdlib_complex128_t *)X;
```
Same comment.

kgryte commented on lib/node_modules/@stdlib/blas/base/zdscal/src/zdscal.c:
```diff
@@ -0,0 +1,35 @@
+* @param X        input array
+* @param strideX  X stride length
+*/
+void API_SUFFIX(c_zdscal)( const CBLAS_INT N, const double alpha, stdlib_complex128_t *X, const CBLAS_INT strideX ) {
```
```suggestion
void API_SUFFIX(c_zdscal)( const CBLAS_INT N, const double alpha, void *X, const CBLAS_INT strideX ) {
```

kgryte commented on lib/node_modules/@stdlib/blas/base/zdscal/src/zdscal_cblas.c:
```diff
@@ -0,0 +1,50 @@
+* @param X        input array
+* @param strideX  X stride length
+*/
+void API_SUFFIX(c_zdscal)( const CBLAS_INT N, const double alpha, stdlib_complex128_t *X, const CBLAS_INT strideX ) {
```
```suggestion
void API_SUFFIX(c_zdscal)( const CBLAS_INT N, const double alpha, void *X, const CBLAS_INT strideX ) {
```

kgryte commented on lib/node_modules/@stdlib/blas/base/zdscal/src/zdscal_cblas.c:
```diff
@@ -0,0 +1,50 @@
+#include "stdlib/blas/base/zdscal.h"
+#include "stdlib/blas/base/zdscal_cblas.h"
+#include "stdlib/blas/base/shared.h"
+#include "stdlib/complex/float64/ctor.h"
```
```suggestion
```

This is also not needed in these header files which use `void *`.

ShabiShett07 replied:
I think we couldn't remove this package from here as we are reinterpreting it

kgryte commented on lib/node_modules/@stdlib/blas/base/zdscal/src/zdscal_cblas.c:
```diff
@@ -0,0 +1,50 @@
+* @param strideX  X stride length
+* @param offsetX  starting index for X
+*/
+void API_SUFFIX(c_zdscal_ndarray)( const CBLAS_INT N, const double alpha, stdlib_complex128_t *X, const CBLAS_INT strideX, const CBLAS_INT offsetX ) {
```
```suggestion
void API_SUFFIX(c_zdscal_ndarray)( const CBLAS_INT N, const double alpha, void *X, const CBLAS_INT strideX, const CBLAS_INT offsetX ) {
```

kgryte commented on lib/node_modules/@stdlib/blas/base/zdscal/src/zdscal_cblas.c:
```diff
@@ -0,0 +1,50 @@
+* @param offsetX  starting index for X
+*/
+void API_SUFFIX(c_zdscal_ndarray)( const CBLAS_INT N, const double alpha, stdlib_complex128_t *X, const CBLAS_INT strideX, const CBLAS_INT offsetX ) {
+	stdlib_complex128_t *x = (stdlib_complex128_t *)X;
```
This is correct.

kgryte replied:
Before adjusting the pointer, we need to reinterpret `X` (which is `void *`) as `stdlib_complex128_t`, which is what you've done here.

kgryte commented on lib/node_modules/@stdlib/blas/base/zdscal/src/zdscal_cblas.c:
```diff
@@ -0,0 +1,50 @@
+void API_SUFFIX(c_zdscal_ndarray)( const CBLAS_INT N, const double alpha, stdlib_complex128_t *X, const CBLAS_INT strideX, const CBLAS_INT offsetX ) {
+	stdlib_complex128_t *x = (stdlib_complex128_t *)X;
+	x += stdlib_strided_min_view_buffer_index( N, strideX, offsetX ); // adjust array pointer
+	API_SUFFIX(cblas_zdscal)( N, alpha, x, strideX );
```
```suggestion
	API_SUFFIX(cblas_zdscal)( N, alpha, (void *)x, strideX );
```

kgryte commented on lib/node_modules/@stdlib/blas/base/zdscal/src/zdscal_f.c:
```diff
@@ -0,0 +1,51 @@
+*/
+void API_SUFFIX(c_zdscal_ndarray)( const CBLAS_INT N, const double alpha, stdlib_complex128_t *X, const CBLAS_INT strideX, const CBLAS_INT offsetX ) {
+	stdlib_complex128_t *x = (stdlib_complex128_t *)X;
+
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/blas/base/zdscal/src/zdscal_f.c:
```diff
@@ -0,0 +1,51 @@
+	stdlib_complex128_t *x = (stdlib_complex128_t *)X;
+
+	x += stdlib_strided_min_view_buffer_index( N, strideX, offsetX );
+	zdscal( &N, alpha, x, &strideX );
```
```suggestion
	zdscal( &N, alpha, (void *)x, &strideX );
```

`void *` elsewhere in this file in the function signatures.

kgryte commented on lib/node_modules/@stdlib/blas/base/zdscal/src/zdscal_ndarray.c:
```diff
@@ -0,0 +1,48 @@
+	if ( N <= 0 || alpha == 1.0 ) {
+		return;
+	}
+	sx = strideX;
```
```suggestion
	x = (stdlib_complex128_t *)X;
	sx = strideX;
```

You need to perform an explicit cast from `void *` to `stdlib_complex128_t`.

kgryte reviewed (changes requested):
Left an initial round of comments.

ShabiShett07 committed:
chore: update implementation

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: missing_dependencies
  - task: lint_c_benchmarks
    status: missing_dependencies
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

ShabiShett07 committed:
chore: update implementation

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

ShabiShett07 committed:
chore: update implementation

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

ShabiShett07 committed:
chore: add void datatype

Co-authored-by: Athan <kgryte@gmail.com>
Signed-off-by: Shabareesh Shetty <139731143+ShabiShett07@users.noreply.github.com>

ShabiShett07 committed:
chore: add void datatype

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

ShabiShett07 committed:
chore: update fortran

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

ShabiShett07 committed:
chore: update pointer

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

ShabiShett07 committed:
chore: update precision

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

ShabiShett07 committed:
chore: update implementation

ShabiShett07 committed:
chore: update indentation

ShabiShett07 committed:
chore: update implementation

Signed-off-by: Shabareesh Shetty <139731143+ShabiShett07@users.noreply.github.com>

ShabiShett07 committed:
chore: update fortan wrapper

ShabiShett07 committed:
chore: update fortan wrapper

ShabiShett07 committed:
remove: unused packages

ShabiShett07 committed:
chore: clean-up

Signed-off-by: Shabareesh Shetty <139731143+ShabiShett07@users.noreply.github.com>

aman-095 commented:
Based on our discussion, make the necessary changes to maintain consistency.

ShabiShett07 committed:
chore: minor clean-up

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

ShabiShett07 committed:
chore: minor clean-up

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

ShabiShett07 committed:
Merge branch 'develop' into feature/c/zdscal

Signed-off-by: Shabareesh Shetty <139731143+ShabiShett07@users.noreply.github.com>

ShabiShett07 committed:
chore: minor clean-up

Signed-off-by: Shabareesh Shetty <139731143+ShabiShett07@users.noreply.github.com>

ShabiShett07 commented:
@aman-095, made necessary changes as per our discussion

kgryte commented on lib/node_modules/@stdlib/blas/base/zdscal/README.md:
```diff
@@ -149,6 +149,151 @@ console.log( x.toString() );
+
+-   **N**: `[in] CBLAS_INT` number of indexed elements.
+-   **alpha**: `[in] double` scalar constant.
+-   **x**: `[inout] stdlib_complex128_t*` input array.
```
```suggestion
-   **X**: `[inout] stdlib_complex128_t*` input array.
```

kgryte committed:
docs: fix param name

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/base/zdscal/README.md:
```diff
@@ -149,6 +149,151 @@ console.log( x.toString() );
+
+-   **N**: `[in] CBLAS_INT` number of indexed elements.
+-   **alpha**: `[in] double` scalar constant.
+-   **x**: `[inout] void*` input array.
```
```suggestion
-   **X**: `[inout] void*` input array.
```

kgryte committed:
docs: fix parameter name

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/base/zdscal/README.md:
````diff
@@ -149,6 +149,151 @@ console.log( x.toString() );
+-   **strideX**: `[in] CBLAS_INT` index increment for `x`.
+
+```c
+void c_zdscal( const CBLAS_INT N, const double alpha, stdlib_complex128_t *X, const CBLAS_INT strideX );
+```
+
+#### c_zdscal_ndarray( N, alpha, \*X, strideX, offsetX )
+
+Scales a double-precision complex floating-point vector by a double-precision floating-point constant using alternative indexing semantics.
+
+```c
+#include "stdlib/complex/float64/ctor.h"
+
+stdlib_complex128_t x[] = {
+    stdlib_complex128( 1.0, 2.0 ),
+    stdlib_complex128( 3.0, 4.0 ),
+    stdlib_complex128( 5.0, 6.0 )
+};
+
+c_zdscal_ndarray( 3, 2.0, x, 1, 0 );
+```
+
+The function accepts the following arguments:
+
+-   **N**: `[in] CBLAS_INT` number of indexed elements.
+-   **alpha**: `[in] double` scalar constant.
+-   **X**: `[inout] void*` input array.
+-   **strideX**: `[in] CBLAS_INT` index increment for `x`.
+-   **offsetX**: `[in] CBLAS_INT` starting index for `x`.
+
+```c
+void c_zdscal_ndarray( const CBLAS_INT N, const double alpha, stdlib_complex128_t *X, const CBLAS_INT strideX, const CBLAS_INT offsetX );
````
```suggestion
void c_zdscal( const CBLAS_INT N, const double alpha, void *X, const CBLAS_INT strideX );
```

#### c_zdscal_ndarray( N, alpha, \*X, strideX, offsetX )

Scales a double-precision complex floating-point vector by a double-precision floating-point constant using alternative indexing semantics.

```c
#include "stdlib/complex/float64/ctor.h"

stdlib_complex128_t x[] = {
    stdlib_complex128( 1.0, 2.0 ),
    stdlib_complex128( 3.0, 4.0 ),
    stdlib_complex128( 5.0, 6.0 )
};

c_zdscal_ndarray( 3, 2.0, x, 1, 0 );
```

The function accepts the following arguments:

-   **N**: `[in] CBLAS_INT` number of indexed elements.
-   **alpha**: `[in] double` scalar constant.
-   **X**: `[inout] void*` input array.
-   **strideX**: `[in] CBLAS_INT` index increment for `x`.
-   **offsetX**: `[in] CBLAS_INT` starting index for `x`.

```c
void c_zdscal_ndarray( const CBLAS_INT N, const double alpha, void *X, const CBLAS_INT strideX, const CBLAS_INT offsetX );
```

kgryte committed:
docs: fix signatures

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
docs: revert changes

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/base/zdscal/README.md:
````diff
@@ -149,6 +149,149 @@ console.log( x.toString() );
+-   **strideX**: `[in] CBLAS_INT` index increment for `x`.
+
+```c
+void c_zdscal( const CBLAS_INT N, const double alpha, stdlib_complex128_t *X, const CBLAS_INT strideX );
````
```suggestion
void c_zdscal( const CBLAS_INT N, const double alpha, void *X, const CBLAS_INT strideX );
```

kgryte committed:
docs: fix type

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/base/zdscal/README.md:
````diff
@@ -149,6 +149,149 @@ console.log( x.toString() );
+-   **strideX**: `[in] CBLAS_INT` index increment for `x`.
+-   **offsetX**: `[in] CBLAS_INT` starting index for `x`.
+```c
+void c_zdscal_ndarray( const CBLAS_INT N, const double alpha, stdlib_complex128_t *X, const CBLAS_INT strideX, const CBLAS_INT offsetX );
+void c_zdscal( const CBLAS_INT N, const double alpha, void *X, const CBLAS_INT strideX );
````
```suggestion
void c_zdscal_ndarray( const CBLAS_INT N, const double alpha, void *X, const CBLAS_INT strideX, const CBLAS_INT offsetX );
```

kgryte committed:
docs: fix signature

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/base/zdscal/README.md:
````diff
@@ -149,6 +149,148 @@ console.log( x.toString() );
+
+c_zdscal_ndarray( 3, 2.0, x, 1, 0 );
+```
+The function accepts the following arguments:
+-   **N**: `[in] CBLAS_INT` number of indexed elements.
+-   **alpha**: `[in] double` scalar constant.
+-   **X**: `[inout] void*` input array.
+-   **strideX**: `[in] CBLAS_INT` index increment for `x`.
+-   **offsetX**: `[in] CBLAS_INT` starting index for `x`.
````
```suggestion

The function accepts the following arguments:
-   **N**: `[in] CBLAS_INT` number of indexed elements.
-   **alpha**: `[in] double` scalar constant.
-   **X**: `[inout] void*` input array.
-   **strideX**: `[in] CBLAS_INT` index increment for `x`.
-   **offsetX**: `[in] CBLAS_INT` starting index for `x`.

```

kgryte committed:
style: fix spacing

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/base/zdscal/README.md:
````diff
@@ -149,6 +149,150 @@ console.log( x.toString() );
+c_zdscal_ndarray( 3, 2.0, x, 1, 0 );
+```
+
+The function accepts the following arguments:
````
```suggestion
The function accepts the following arguments:

```

kgryte committed:
docs: fix spacing

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/base/zdscal/README.md:
```diff
@@ -149,6 +149,151 @@ console.log( x.toString() );
+#include <stdio.h>
+
+int main( void ) {
+    // Create strided arrays:
```
```suggestion
```

kgryte committed:
docs: remove comment

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/base/zdscal/README.md:
```diff
@@ -149,6 +149,150 @@ console.log( x.toString() );
+    // Specify the number of elements:
+    const int N = 4;
+
+    // Specify stride lengths:
```
```suggestion
    // Specify the stride length:
```

kgryte committed:
docs: fix comment

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/base/zdscal/benchmark/c/benchmark.length.c:
```diff
@@ -0,0 +1,191 @@
+	}
+	t = tic();
+	for ( i = 0; i < iterations; i++ ) {
+		c_zdscal( len, 2.0, x, 1 );
```
```suggestion
		c_zdscal( len, 2.0, (void *)x, 1 );
```

kgryte committed:
bench: add cast

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/base/zdscal/benchmark/c/benchmark.length.c:
```diff
@@ -0,0 +1,191 @@
+	}
+	t = tic();
+	for ( i = 0; i < iterations; i++ ) {
+		c_zdscal_ndarray( len, 2.0, x, 1, 0 );
```
```suggestion
		c_zdscal_ndarray( len, 2.0, (void *)x, 1, 0 );
```

kgryte committed:
bench: add cast

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/base/zdscal/examples/c/example.c:
```diff
@@ -0,0 +1,53 @@
+	// Specify the number of elements:
+	const int N = 4;
+
+	// Specify stride lengths:
```
```suggestion
	stdlib_complex128_t x[] = {
		stdlib_complex128( 1.0, 2.0 ),
		stdlib_complex128( 3.0, 4.0 ),
		stdlib_complex128( 5.0, 6.0 ),
		stdlib_complex128( 7.0, 8.0 )
	};

	// Specify the number of elements:
	const int N = 4;

	// Specify the stride length:
```

kgryte committed:
docs: update comments

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
refactor: update include guards

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
refactor: update include guards

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
refactor: update include guards and fix include

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/base/zdscal/include/stdlib/blas/base/zdscal_fortran.h:
```diff
@@ -0,0 +1,43 @@
+/**
+* Scales a double-precision complex floating-point vector by a double-precision floating-point constant.
+*/
+void zdscal( const CBLAS_INT *, const double *, stdlib_complex128_t *, const CBLAS_INT * );
```
```suggestion
void zdscal( const int *, const double *, void *, const int * );
```

kgryte committed:
fix: use correct signature

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/base/zdscal/include/stdlib/blas/base/zdscal_fortran.h:
```diff
@@ -0,0 +1,43 @@
+#ifndef STDLIB_BLAS_BASE_ZDSCAL_FORTRAN_H
+#define STDLIB_BLAS_BASE_ZDSCAL_FORTRAN_H
+
+#include "stdlib/complex/float64/ctor.h"
+
```
```suggestion
```

kgryte committed:
refactor: remove include

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/base/zdscal/include/stdlib/blas/base/zdscal.h:
```diff
@@ -0,0 +1,49 @@
+#ifndef STDLIB_BLAS_BASE_ZDSCAL_H
+#define STDLIB_BLAS_BASE_ZDSCAL_H
+
+#include "stdlib/blas/base/shared.h"
+#include "stdlib/complex/float64/ctor.h"
```
```suggestion
#include "stdlib/blas/base/shared.h"
```

kgryte committed:
refactor: remove include

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/base/zdscal/include/stdlib/blas/base/zdscal_cblas.h:
```diff
@@ -0,0 +1,44 @@
+#ifndef STDLIB_BLAS_BASE_ZDSCAL_CBLAS_H
+#define STDLIB_BLAS_BASE_ZDSCAL_CBLAS_H
+
+#include "stdlib/blas/base/shared.h"
+#include "stdlib/complex/float64/ctor.h"
```
```suggestion
#include "stdlib/blas/base/shared.h"
```

kgryte committed:
refactor: remove include

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/base/zdscal/src/zdscal.f:
```diff
@@ -0,0 +1,80 @@
+  ! ..
+  if ( strideX == 1 ) then
+    do i = 1, N
+      x( i ) = cmplx(alpha*dble(x(i)),alpha*aimag(x(i)), kind=kind(0.0d0))
+    end do
+  else
+    nstrideX = N * strideX;
+    do i = 1, nstrideX, strideX
+      x( i ) = cmplx(alpha*dble(x(i)),alpha*aimag(x(i)), kind=kind(0.0d0))
```
```suggestion
      x( i ) = cmplx( alpha*dble(x(i)), alpha*aimag(x(i)), kind=kind(0.0d0) )
    end do
  else
    nstrideX = N * strideX;
    do i = 1, nstrideX, strideX
      x( i ) = cmplx( alpha*dble(x(i)), alpha*aimag(x(i)), kind=kind(0.0d0) )
```

kgryte committed:
style: add spaces

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
refactor: remove unnecessary variable

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (approved):
LGTM after clean-up.
